src/webutil/tableState.js

```diff
@@ -0,0 +1,185 @@
+   */
+  +rowsPerPage: number,
+  /**
+    A stateful representation of the current sort name (the last sort name passed to setSortFn).
+   */
+  +sortName: string,
+  /**
+    A stateful representation of the current order, represented by "asc" or "desc".
+   */
+  +sortOrder: SortOrderTypes,
+  /**
+    A function that changes the currentPage to represent the indicated index.
+   */
+  +setPageIndex: (number) => void,
+  /**
+    A function that sets the rows per page and updates the pageIndex to an equivalent place.
+   */
+  +setRowsPerPage: (number) => void,
+  /**
+    A function that adds a filter to the map of filters or overwrites an existing filter that matches the key.
+    Every filter stored will be applied together as an AND operation (results will only include elements
+    that pass all filters).
+    Sets pageIndex to 0.
+
+    string: A key that represents the filter in the map.
+    function?: A function that returns false if an element should be excluded from the result. Null to remove the filter at the key.
+   */
+  +createOrUpdateFilterFn: (string, null | ((T) => boolean)) => void,
+  /**
+    A function that sorts the data. Current pageIndex is maintained.
+
+    string: A key representing the sort field. It is is used to toggle sort order and increase efficiency.
+    function?: A function that returns the property of the element that should be used in the sort comparator.
+   */
+  +setSortFn: (string, PluckFn<T>) => void,
+  /**
+    A function that overwrites the base data and reapplies current sort and filter functions.
+    Sets pageIndex to 0.
+   */
+  +setData: ($ReadOnlyArray<T>) => void,
+|};
+
+/**
+  A builder function for initializing a new TableState for use in a ReactNode.
+  It encapsulates Sort, Pagination, and Filtering.
+ */
+export function useTableState<T>(defaults?: {
+  initialRowsPerPage?: number,
+  initialSort?: {
+    // The starting order of the initial sort, represented by "asc" or "desc".
+    sortOrder: SortOrderTypes,
+    // A key representing the sort field. It is is used to toggle sort order and increase efficiency.
+    sortName: string,
+    // A function that returns the property of the element that should be used in the sort comparator.
+    sortFn: PluckFn<T>,
+  },
+}): TableState<T> {
+  defaults = defaults || {};
+  const sortDefaults = defaults.initialSort || {};
+  const [originalArray, setOriginalArray] = useState([]);
+  const [filteredArray, setFilteredArray] = useState([]);
+  const [sortName, setSortName] = useState(sortDefaults.sortName || "");
+  const [sortOrder, setSortOrder] = useState(
+    sortDefaults.sortOrder || SortOrders.ASC
+  );
+  const [pageIndex, setPageIndex] = useState(0);
+  const [rowsPerPage, setRowsPerPageInternal] = useState(
+    defaults.initialRowsPerPage || 20
+  );
+  let _sortFn: PluckFn<T> | null = sortDefaults.sortFn || null;
+  const _filterFns: Map<string, null | ((T) => boolean)> = new Map();
```

Comment:
This Map is outside of React's state management, so its likely to cause bugs / issues with re-rendering components when the filter functions change. Its best to keep everything within Reacts state management. Additionally, Map isn't a good fit because when you call `.set()`, the data inside changes but the Map itself remains referentially equal to its previous state, which means React doesn't detect any changes and therefore doesn't re-render the UI.

A better way to do this would be to also put this in useState:
```ts
type FilterFnState = { [key: string]: null | ((T) => boolean) };

// ...
const [filterFns, setFilterFns] = useState<FilterFnState>({});

const createOrUpdateFilterFn = (
     key: string,
     newFilterFn: $Values<FilterFnState>
   ) => {
     setPageIndex(0);
     // if you pass in a function instead of a value, you can get the current filter state as a param and return the new state
     setFilterFns(oldFilters => ({
       ...oldFilters,
       [key]: newFilterFn,
     }));
     // no need to do .slice since .filter returns a new array instead of mutating the input array
     _filter(originalArray);
   };
```

Reply:
Again, don't really get the nuance, but I'll play with this.

Reply:
accepted